app/src/main/java/com/ncnf/map/MapHandler.java

```diff
@@ -90,24 +98,40 @@ public LatLng getUserPosition(){
     }
 
     private void addEventMarkers(){
-        List<Event> events = eventDB.toList();
-        eventMarkers = new ArrayList<>();
+        List<Event> events = queryEvents();
+        Map<LatLng, List<Event>> eventMap = new HashMap<>();
         for (Event p : events) {
             LatLng event_position = new LatLng(p.getLocation().getLatitude(), p.getLocation().getLongitude());
             if (MapUtilities.position_in_range(event_position, userPosition)){
-                clusterManager.addItem(new com.ncnf.map.Marker(event_position, p.getName(), p.getDescription()));
+                if (!eventMap.containsKey(event_position)){
+                    eventMap.put(event_position, new ArrayList<>());
+                }
+                eventMap.get(event_position).add(p);
             }
         }
+        Set<LatLng> keys = eventMap.keySet();
+        for (LatLng k : keys){
+            List<Event> list = eventMap.get(k);
+            StringBuilder desc = new StringBuilder();
+            for (Event p : list){
+                desc.append(p.getName()).append("\n");
+            }
+            String description = desc.toString();
+            clusterManager.addItem(new com.ncnf.map.Marker(k, description, eventMap.get(k).get(0).getAddress(), list));
+        }
     }
 
     private void addVenueMarkers(){
         List<Venue> venues = venueProvider.getAll();
-        venueMarkers = new ArrayList<>();
         for (Venue p : venues) {
             LatLng venue_position = new LatLng(p.getLatitude(), p.getLongitude());
             if (MapUtilities.position_in_range(venue_position, userPosition)){
-                clusterManager.addItem(new com.ncnf.map.Marker(venue_position, p.getName(), p.getName()));
+                clusterManager.addItem(new com.ncnf.map.Marker(venue_position, p.getName(), p.getName(), null));
```

Comment:
You can make a static import of the location to have only Marker() without needing to specify the path